### PR TITLE
Add BoxedType GenericType

### PR DESCRIPTION
**Pull request by devinrsmith**
This increases the expressiveness of Type, allowing us to capture the differences between primitives and their boxed counterparts.

### CODE REVIEW

File: qst/src/main/java/io/deephaven/qst/type/BoxedType.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ */
+package io.deephaven.qst.type;
+
+import io.deephaven.annotations.SimpleStyle;
+import org.immutables.value.Value.Immutable;
+import org.immutables.value.Value.Parameter;
+
+@Immutable
+@SimpleStyle
+public abstract class BoxedType<T> extends GenericTypeBase<T> {
```
[the remark this answers is not in the document]
Comment (devinrsmith) [reply qst/src/main/java/io/deephaven/qst/type/BoxedType.java]: 8, I think.

File: qst/src/main/java/io/deephaven/qst/type/BoxedTypeVisitorAdapter.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    private final PrimitiveType.Visitor<R> delegate;
+
+    public BoxedTypeVisitorAdapter(Visitor<R> delegate) {
+        this.delegate = Objects.requireNonNull(delegate);
+    }
+
+    @Override
+    public R visit(BoxedBooleanType booleanType) {
+        return delegate.visit(booleanType.primitiveType());
+    }
```
Comment (rcaudy) [inline qst/src/main/java/io/deephaven/qst/type/BoxedTypeVisitorAdapter.java:17]: Do we need this much code? Why not just walk the delegate from the BoxedType's primitive?

File: qst/src/main/java/io/deephaven/qst/type/BoxedByteType.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.immutables.value.Value.Immutable;
+
+/**
+ * The {@link Byte} type.
+ */
+@Immutable
+@SingletonStyle
+public abstract class BoxedByteType extends BoxedTypeBase<Byte> {
+
+    public static BoxedByteType of() {
```
Comment (rcaudy) [inline qst/src/main/java/io/deephaven/qst/type/BoxedByteType.java:16]: We called this `instance()` on `PrimitiveType` implementations. Should we standardize?

File: qst/src/main/java/io/deephaven/qst/type/PrimitiveTypeVisitorAdapter.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+class PrimitiveTypeVisitorAdapter<R> implements PrimitiveType.Visitor<R> {
+    private final BoxedType.Visitor<R> delegate;
+
+    public PrimitiveTypeVisitorAdapter(Visitor<R> delegate) {
+        this.delegate = Objects.requireNonNull(delegate);
+    }
+
+    @Override
+    public R visit(BooleanType booleanType) {
+        return delegate.visit(booleanType.boxedType());
```
Comment (rcaudy) [inline qst/src/main/java/io/deephaven/qst/type/PrimitiveTypeVisitorAdapter.java:16]: Same question. Can't we just walk the primitive?

File: qst/src/main/java/io/deephaven/qst/type/BoxedTypeVisitorAdapter.java (hunk 2 above)
Comment (devinrsmith) [reply qst/src/main/java/io/deephaven/qst/type/BoxedTypeVisitorAdapter.java:17]: Yes - that's a good pattern, and cuts down on the code. I thought I would need these adapters, but ultimately didn't (and used the pattern like you suggested).

File: qst/src/main/java/io/deephaven/qst/type/BoxedByteType.java (hunk 3 above)
Comment (devinrsmith) [reply qst/src/main/java/io/deephaven/qst/type/BoxedByteType.java:16]: Updated to `of()`.

**Review by rcaudy — APPROVED** [review]

**Pull request merged**